examples/hybrid_parallelism/run_pretraining.py

```diff
@@ -0,0 +1,504 @@
+            total_loss.persistable = True
+            # checkpoints.append(total_loss.name)
+
+    if args.use_sop:
+        graph_vars = {
+            'data_loader': data_loader,
+            'mask_lm_loss': mask_lm_loss,
+            'mean_mask_lm_loss': mean_mask_lm_loss,
+            'sop_loss': mean_sop_loss,
+            'sop_acc': sop_acc,
+            'total_loss': total_loss,
+            'checkpoints': checkpoints
+        }
+    else:
+        graph_vars = {
+            'data_loader': data_loader,
+            'mask_lm_loss': mask_lm_loss,
+            'mean_mask_lm_loss': mean_mask_lm_loss,
+            'total_loss': total_loss,
+            'checkpoints': checkpoints,
+        }
+    return graph_vars
+
+
+def train(args):
+    log.info("pretraining start")
+    profile = False
+
+    place = fluid.CUDAPlace(int(os.environ.get('FLAGS_selected_gpus', 0)))
+
+    # set seed
+    random.seed(args.seed)
+    np.random.seed(args.seed)
+    paddle.seed(args.seed)
+    get_rng_state_tracker().add('global_seed', args.seed)
+    get_rng_state_tracker().add('local_seed', args.seed + fleet.worker_index() + 2021)
+
+    # define execution strategy
+    exec_strategy = fluid.ExecutionStrategy()
+    exec_strategy.num_threads = 2
+    exec_strategy.num_iteration_per_drop_scope = 1
+
+    # define distribution strategy
+    dist_strategy = fleet.DistributedStrategy()
+    dist_strategy.execution_strategy = exec_strategy
+    dist_strategy.nccl_comm_num = 3
+    if args.use_recompute:
+       log.info("using recompute.")
+    dist_strategy.recompute = args.use_recompute
+    dist_strategy.sharding = args.use_sharding
+    dist_strategy.pipeline = args.num_pp > 1
+
+    # define topology structure for dp/pp/mp
+    topo = Topology(rank=fleet.worker_index(),
+                    world_size=fleet.worker_num(),
+                    dp=args.num_dp,
+                    pp=args.num_pp,
+                    sharding=args.num_sharding,
+                    mp=args.num_mp)
+
+    is_last = False
+    if topo.pp.rank ==  (topo.pp.size -1) :
+        is_last = True
+
+    dp_sharding_rank = topo.dp.rank * topo.sharding.size + topo.sharding.rank
+    dp_worldsize = topo.dp.size * topo.sharding.size
+    bsz_per_dp = args.global_bsz // dp_worldsize
+
+    micro_bsz = args.micro_bsz
+    assert args.global_bsz % micro_bsz == 0, f"cannot do gradient accumulate, globa_bsz: {args.bsz} micro_bsz: {micro_bsz}"
+    acc_steps = bsz_per_dp // micro_bsz
+
+    # sharding \ model parallel \ pipeline
+    if args.use_sharding:
+        dist_strategy.sharding_configs = {"segment_broadcast_MB": 32,
+                                           "sharding_degree": args.num_sharding,
+                                           "mp_degree": args.num_mp,
+                                           "pp_degree": args.num_pp,
+                                           "dp_degree":args.num_dp,
+                                           "hybrid_dp": False,
```

Comment:
hybrid_dp setting is deprecate and should be remove

Reply:
Done.